xarray/core/common.py

```diff
@@ -576,7 +576,13 @@ def where(self, cond, other=None, drop=False):
             outobj = self
             outcond = cond
 
-        return outobj._where(outcond)
+        # preserve attributes
+        out = outobj._where(outcond)
+        out.attrs = self.attrs
+        if isinstance(out, Dataset):
+            for v in out:
+                out[v].attrs = self[v].attrs
```

Comment:
Use `out.variables[v].attrs = self.variables[v].attrs` instead. Creating `DataArray` objects is pretty expensive compared to just pulling out existing `Variable` objects.